gamla/functional_generic.py

```diff
@@ -539,3 +539,22 @@ def merge_with(*dicts):
 merge = merge_with(functional.last)
 concat = itertools.chain.from_iterable
 mapcat = compose_left(curried_map, after(concat))
+
+
+def groupby(key: Callable):
+    return compose_left(
+        functional.groupby_many_reduce(
+            compose_left(key, functional.wrap_tuple),
+            compose_left(
+                functional.pack,
+                stack(
+                    [
+                        unless(functional.identity, functional.just(())),
+                        functional.wrap_tuple,
+                    ]
+                ),
+                concat,
+            ),
+        ),
+        valmap(frozenset),
```

Comment:
you sure we want a set? what if we count on it?

Reply:
I just conformed to graph.groupby_many. 
But you are right. So tuple? leave it as a chain?

Reply:
yes i'm not sure myself
i would say a tuple, because we're not really saving memory by storing it as a generator, but we can easily introduce bugs

Reply:
yup, going for tuple